### DISCUSSION TITLE
Allow queryset for RelatedField to be a callable

### DISCUSSION THREAD

**Discussion by tuky** [encode/django-rest-framework#8831] [Ideas & Suggestions]
I would like to suggest this to make 2 use cases easier:

1. The callable could optionally receive the `context` so you can for example filter on the authenticated user without overriding each such field (see https://github.com/encode/django-rest-framework/issues/1985)
2. We have querysets like `Package.objects.filter(expiration__lt=timezone.now())` which we really need to be lazy.

What do you think? Thx!

**Comment by alextatarinov** [comment]
I agree this would be a useful feature. 
I feel like a better option could be similar to custom field validation, but instead of `def validate_field_name(self, value)` it can be `def get_queryset_field_name(self, field)`. Having access to `self` is quite useful and the following typical pattern definitely won't fit into lambda:
```
def get_fields(self):
    fields = super().get_fields()

    if getattr(self.context['view'], 'swagger_fake_view', False):
        queryset = Model.objects.none()
    else:
        user = self.context['request'].user
        queryset = Model.objects.filter(owner=user)

    fields['field_name'].queryset = queryset
    return fields
```

**Reply by tuky** [reply]
great idea. it would be also similar to how `SerializerMethodField` works. maybe we'd need a `queryset_method_name` parameter, too.
